### ISSUE TITLE
Make '(' and ')' optional in statements, make ';' optional at the end of lines

### ISSUE THREAD

**Issue by dmitryuck** [dart-lang/language#1264]
How do you like this idea?

Make '(' and ')' optional in statements,

Also, make ';' optional at the end of lines

Would be like swift and golang modern languages

**Comment by lrhn** [comment]
The `;` is already covered by #69 and #72. It's not looking like it's going to be an easy change to make.

For omitting the parentheses around the expressions of `while`/`do`-`while`/`for`/`switch`/`if`, it's *probably* possible.
It can will parsing harder, though, and will introduce some ambiguities in the grammar..

Take `if (test) [e1].forEach(action);` which is currently a valid Dart `if` expression. If we omitted the parentheses, the parser would have to work hard to avoid parsing `if test [e1].forEach(action);` as `if (test[e1].forEach(action));` 
That particular code is unlikely, but it means that we have introduced ambiguity into the grammar, and we'll have to find some way to resolve that in general, even for weird programs that might not happen in practice (making the distinction between which programs won't happen in practice and which will is probably at least as hard as solving the problem in general).

We could say that you can only omit the parentheses if you use braces. Then you'd have to choose between
```dart
if (test) [e1].forEach(action);
// or
if test {
  [e1].forEach(action);
}
```
(and I don't think the formatter would allow the latter to stay on one line).

The `switch` cannot omit braces anyway,  so that one should be safe (as long as we don't introduce suffix `{something}` syntax for expressions. Which we might, say for trailing function arguments.

**Comment by lrhn** [comment]
Yes,, that is indeed part of the problem. We can add heuristics, for example whitespace, to solve ambiguities, but it means that dartfmt can no longer safely change whitespace. Or, it has to pick an interpretation first *without* using the whitespace, and then introduce the whitespace to match. Either is troublesome.

**Comment by awkj** [comment]
I was forced to learn Dart in order to learn flutter. Although Dart was released in 2011, compared to kotlin swift, if I hadn't been told, I would have thought it was released in 1991 and had none of the features that make modern languages so easy to write, type post, omitting semicolons, etc. Flutter is good, but Dart is really bad.

**Comment by gosoccerboy5** [comment]
```dart
if (foo) { }
{
  print("hello");
}
```
would be ambiguous, right?

**Comment by lrhn** [comment]
That too, yes. We can probably get around that by disallowing function literals in condition positions, because they can't possibly have type `bool` anyway, but it's a good example of why grammar without delimiters is hard to parse. Being adjacent sometimes means something, and we use the required separators and delimiters with *known* meaning as guides to know whether things belong together or apart.

**Comment by gosoccerboy5** [comment]
Oh, good point

**Comment by Hixie** [comment]
`if (foo) { return foo; } (foo) { }`

**Comment by munificent** [comment]
I definitely don't see us making the parentheses in statement headers optional.
